### DISCUSSION TITLE
3-d probing w. iosender - making a height map

### DISCUSSION THREAD

**Discussion by Klippert** [terjeio/grblHAL#286] [Q&A]
Hello everybody,

I'm getting started on the 3d probe. So first of all, i need to understand how to really use the 3d probe. 

Here's what i'm thinking, please correct me where i'm wrong.
as the 3dprobe goes in the spindle, in a collet, i have no way of knowing how much it's sticking out of the collet, from time to time i use it. For x and y probing, this doesnt' matter, as long as the probe ball is hitting the object in the right place i'm good.

For Z, this means that i can only probe relatively, like probing the baseplate and then probing the thing i want to know the height of should give me the difference between the baseplate and the object i'm probing. 

If I can probe my fixture probe with (without triggering the probe first) i could use that to actually "set" my probe height, but i guess i'm relying on the springs in the probe and the fixture to be different. 

First off I'm trying to figure out how to do a height map of my baseplate. Just to get a sense of how much i'm off, and how good or bad my facing actually is. 

what's the best way of doing such a probing pattern, where i'd probe every 25mm in a grid, and get a bit of a map of the baseplate?
Does IO-sender support any such probing? (or should it be set up in Fusion?)

Any help would be greatly appreciated.

**Comment by phil-barrett** [comment]
You define an area to probe and a grid size.  The way I use it:

- Load the GCode program
- Make a height map with the 3D probe (use set from program limits and a reasonably small grid size)
- Change to the bit I'm going to route with
- Zero Z at the height map's 0, 0 location
- Apply the height map (apply button on the probe panel)
- Cycle Start

**Reply by Klippert** [reply]
So would that mean that if your baseplate is not entirely flat, and you do a heightmap, and then put a piece of material on the plate strapping it down firmly, so it follows the baseplate, and face it, once you remove it from the baseplate, it would be flat?

I got to experiement a bit with that.

**Reply by terjeio** [reply]
> First off I'm trying to figure out how to do a height map of my baseplate. Just to get a sense of how much i'm off, and how good or bad my facing actually is.

If you save the height map to a file you can then open it in a text editor or a web browser to see the measurement results.

> So would that mean that if your baseplate is not entirely flat, and you do a heightmap, and then put a piece of material on the plate strapping it down firmly, so it follows the baseplate, and face it, once you remove it from the baseplate, it would be flat?

I would not count on that. If the base plate is uneven how can you ensure the workpiece will follow it exactly?

**Reply by Klippert** [reply]
ok. did an experiment, and in the name of experiments it went:
![image](https://user-images.githubusercontent.com/53910032/113925862-c7324000-97eb-11eb-9e97-98278b48ac44.png)

so when doing a height map, it does not seem to retract between the measurements (not enough, anyhow) - as i was doing a test heightmap and had left a 10mm high clamp on the build plate. it did a rapid move to the next probe point, without retracting high enough to clear the clamp. 

So how do you set the retract height between probings?

also i noticed that when you start probing for a height map, if one selects probe from x,y 0,0, it does a rapid move to 0,0 without retracting Z first. so that almost cost me probe pin #2. 

so, now i'm more careful. and still learning.

@terjeio - what do you use the height map for, if not to compensate Z while running? so you have the same height over your build plate at all times? or am i missing something here?

**Reply by phil-barrett** [reply]
It has been a while since I used it and, iirc, it was with ioSender Beta 8.  

I think what Terje was saying is there is no guarantee that the workpiece will conform to your spoil board. This makes sense in general though the actual deviation may be small in most cases. If you have surfaced your spoil board then it is likely to be fairly close.  For things like etching where the reference is the original surface of the workpiece, I prefer to make the height map against the actual workpiece. Then, there is no question of it's correctness. 

And, by the way, if you are etching PCBs, be aware that the material may be bowed.  I use double sided tape to ensure that it is as close to flat against the spoil board as possible.

**Reply by Klippert** [reply]
That does make good sense as a use case for a height map. 

I want to use it to check my spoilboard flatness and to check the tolerances of my parts. I know this is a bit of a no-no but it’s my best bet aside from just measuring analog points. 

I’m building a new z-axis and have two pieces that has to fit with Lino ear guides and ball screws, which doesn’t like too big tolerances. 

Is there a way to set the S-height it retracts to between probing?

**Reply by terjeio** [reply]
> Is there a way to set the S-height it retracts to between probing?

Yes and no - it is a bug as I use the _Depth_ from _Probing clearances_ for the retract. This is IMO wrong as the _Depth_ field is disabled (as it should be), I will fix this by retracting to the initial Z position instead.

A new [edge version](http://www.io-engineering.com/downloads) has been uploaded with this fix.

**Reply by phil-barrett** [reply]
I ran a height map on my spoilboard before and after leveling. Was mostly curious. Before leveling, it showed about a 1.5mm deviation across the entire spoilboard.  It read within .1 mm after leveling.  Not perfect but acceptable for my needs.  

Note that this does not necessarily mean my spoilboard is flat within .1mm - my X and Y rails could easily be off as well. If so, the milled spoilboard conformed to that.

**Reply by Klippert** [reply]
> > Is there a way to set the S-height it retracts to between probing?
> 
> Yes and no - it is a bug as I use the _Depth_ from _Probing clearances_ for the retract. This is IMO wrong as the _Depth_ field is disabled (as it should be), I will fix this by retracting to the initial Z position instead.
> 
> A new [edge version](http://www.io-engineering.com/downloads) has been uploaded with this fix.

Excellent. I'll test it in the weekend. - Is the fix committed to the IOSender on github? or the new repository?

**Reply by terjeio** [reply]
> Is the fix committed to the IOSender on github?

Not yet - a pre built binary can be found [here](http://www.io-engineering.com/downloads).

> or the new repository?

The sender is not going to be moved to the new repo, it will stay where it is and renamed to ioSender on the first production release which will be soon.

**Reply by Klippert** [reply]
Thanks - i'll try out the new version. 
Ok. this is much better, if i set a reasonably safe Z height it correctly pulls up to that height before moving to next probe point. 
I successfully did a heightmap of my buildplate. And nothing is ever flat ;)

I did notice a few mishaps. 
My probe sometimes "locks up" - so it dosn't let go, despite being pulled up or away from what's being probed. 
this causes an alarm 4. - which seems to lock up io-sender. I can click on the settings tab and the app settings, but none of the other tabs want to come to life. the top of the tab just goes slightly purple, and nothing happens. some times io-sender crashes and other times i can switch back and forth between random tabs and wait a little and then it suddenly appears (the grbl tab) where i can unlock and continue life. 
Other times i just have to restart. happy path works, but if i encounter an error during probing, it's pretty certain that restarting iosender is the only way forward.

I've no idea why. 

Also fusion has a probing function that i'm quite keen on, alignment of an object. i.e. probe x at one point and then move 50mm y+ and probe x again, and let me know if my piece is straight or crooked in on the baseplate/in the vice. 
Is that something that would be complex to add?

**Comment by Klippert** [comment]
is there a way of getting a graphical representation (heatmap) of a
heightmap?

**Reply by terjeio** [reply]
You mean like this?

![image](https://user-images.githubusercontent.com/20260062/114150939-e41a6080-991c-11eb-9713-492f9b00c0af.png)

It is on my todo list to overlay the heatmap on the gcode 3d view.

**Reply by Klippert** [reply]
yes - exactly. though some measurement legend would be great (ie. blue +0.1mm red -0.2mm or whatever, just so i would know what i'm looking at.)

And the idea of overlaying it the 3d view of the gcode would be so amazing! 
So i could essentially probe a whole part, and then overlay the result to the model, and see how my part matches specs. - Perhaps one could even set up a tolerance say 0,05mm, and coclor code, where am i within specs and where do i need to continue working to get my part right.

**Reply by phil-barrett** [reply]
I think the colors are unitless - blue is height map min, red is max.

**Reply by Klippert** [reply]
It would be helpfull to me to at least know the Z-min and Z-max - is ther an easy way of getting those values?

**Reply by Klippert** [reply]
maybe a Z-avg. or even a std. deviation?

**Reply by terjeio** [reply]
> My probe sometimes "locks up" - so it dosn't let go, despite being pulled up or away from what's being probed.
this causes an alarm 4. - which seems to lock up io-sender. 

I'll have to check this.

> I think the colors are unitless - blue is height map min, red is max.

Correct.

> It would be helpfull to me to at least know the Z-min and Z-max - is ther an easy way of getting those values?

If you save the map you can find the z-values in that. It is in xml-format, if you drop it into Excel you will find the Z offsets in the _point_ column. Or you can open the file in a text editor.

> maybe a Z-avg. or even a std. deviation?

I'll add the deviation to the status line when probing is completed. 

> Also fusion has a probing function that i'm quite keen on, alignment of an object. i.e. probe x at one point and then move 50mm y+ and probe x again, and let me know if my piece is straight or crooked in on the baseplate/in the vice.
Is that something that would be complex to add?

This is on my todo list, another gcode transform. It will not be added to the first production release though.

**Reply by terjeio** [reply]
>>    maybe a Z-avg. or even a std. deviation?

> I'll add the deviation to the status line when probing is completed.

It is already in the status line. Or do you want something else?

![image](https://user-images.githubusercontent.com/20260062/114317590-14beed80-9b09-11eb-847c-e75fe6eb95ac.png)

**Reply by Klippert** [reply]
I didnt notice it in the status Window. Ill check. It might be good to have it as a pop-up or in the Console, so it doesn’t dissappear then one absentmindedly starts jogging or something like that.
